MayssaraOmar/assignment2/Assignment 2.cpp

```diff
@@ -0,0 +1,275 @@
+		const T &data) {
+	if (current) {
+		if (!current->left) {
+			current->left = new tree_node<T>(data, nullptr, nullptr);
+		} else {
+			current->left->data = data;
+		}
+		return current->left;
+	}
+	return nullptr;
+}
+template<class T>
+tree_node<T>* binary_tree<T>::insert_right(tree_node<T>* current,
+		const T &data) {
+	if (current) {
+		if (!current->right) {
+			current->right = new tree_node<T>(data, nullptr, nullptr);
+
+		} else {
+			current->right->data = data;
+		}
+		return current->right;
+	}
+	return nullptr;
+}
+template<class T>
+tree_node<T>* binary_tree<T>::get_root() {
+	return root;
+}
+template<class T>
+tree_node<T>* binary_tree<T>::get_left(tree_node<T>* current) {
+	if (current) {
+		return current->left;
+	}
+	return nullptr;
+}
+template<class T>
+tree_node<T>* binary_tree<T>::get_right(tree_node<T>* current) {
+	if (current) {
+		return current->right;
+	}
+	return nullptr;
+}
+template<class T>
+void binary_tree<T>::remove_leaf(tree_node<T>* current) {
+	if (current && !current->left && !current->right) {
+		delete current;
+	}
+}
+
+template<class T>
+bool binary_tree<T>::empty() {
+	return root == nullptr;
+}
+template<class T>
+void binary_tree<T>::clear() {
+	if (empty()) {
+		return;
+	}
+	stack<tree_node<T>*> st;
+	st.push(root);
+	while (!st.empty()) {
+		tree_node<T>* node = st.top();
+		st.pop();
+		if (node->right)
+			st.push(node->right);
+		if (node->left)
+			st.push(node->left);
+		delete node;
+	}
+	root = nullptr;
+	return;
+}
+template<class T>
+bool binary_tree<T>::print_ancestors_helper(tree_node<T>* current,
+		const T &target) {
+	if (!current) {
+		return false;
+	}
+
+	if (print_ancestors_helper(current->left, target)) {
+		cout << current->data << ' ';
+		return true;
+	}
+	if (print_ancestors_helper(current->right, target)) {
+		cout << current->data << ' ';
+		return true;
+	}
+	if (current->data == target) {
+		return true;
+	}
+	return false;
+}
+template<class T>
+void binary_tree<T>::print_ancestors(const T &target) {
+	if (!print_ancestors_helper(root, target)) {
+		cout << "ERROR: key not found in tree";
+	}
+	cout << endl;
+}
+template<class T>
+pair<tree_node<T>*, int> binary_tree<T>::common_ancestor_helper(
+		tree_node<T>* current, tree_node<T>* first, tree_node<T>* second) {
+
+	if (!current || !first || !second || first == second) {
+		return make_pair(nullptr, 0);
+	}
+
+	pair<tree_node<T>*, int> res_left = common_ancestor_helper(current->left,
+			first, second);
+	if (res_left.first)
+		return res_left;
+
+	pair<tree_node<T>*, int> res_right = common_ancestor_helper(current->right,
+			first, second);
+	if (res_right.first)
+		return res_right;
+
+	if (current == first || current == second)
+		res_left.second++;
+	res_left.second = res_left.second + res_right.second;
+	if (res_left.second == 2) {
+		res_left.first = current;
+		return res_left;
+	}
+
+	return res_left;
+}
+template<class T>
+pair<tree_node<T>*, T> binary_tree<T>::common_ancestor(tree_node<T> *first,
```

Comment:
Returning a pair here is not needed, right?
You could just return either a `tree_node<T>*` or a `T *`